modules/host/contractmanager/storagefolderadd.go

```diff
@@ -6,9 +6,42 @@ import (
 	"sync/atomic"
 
 	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/crypto"
 	"github.com/NebulousLabs/Sia/modules"
 )
 
+// findUnfinishedStorageFolderAdditions will scroll through a set of state
+// changes and figure out which of the unfinished storage folder additions are
+// still unfinished. If a storage folder addition has finished, it will be
+// discoverable through the presence of a storage folder addition object using
+// the same path, or through an errored storage folder addition object using
+// the same path. Putting the logic here keeps the WAL logic cleaner.
+func findUnfinishedStorageFolderAdditions(scs []stateChange) []*storageFolder {
+	// Create a map containing a list of all unfinished storage folder
+	// additions identified by their path to the storage folder that is being
+	// added.
+	usfMap := make(map[uint16]*storageFolder)
+	for _, sc := range scs {
+		for _, sf := range sc.UnfinishedStorageFolderAdditions {
+			usfMap[sf.Index] = sf
+		}
+		for _, sf := range sc.StorageFolderAdditions {
+			delete(usfMap, sf.Index)
+		}
```

Comment:
If there's an addition, it'll appear in a strictly later scs element than the corresponding 'unfinished'. The state change array is like a set of transactions in a block - you process them in order and they can depend on parents, but parents can't depend on children.